src/audio/tdfb/tdfb_generic.c

```diff
@@ -81,7 +81,9 @@ void tdfb_fir_s16(struct tdfb_comp_data *cd, struct input_stream_buffer *bsource
 			/* Read two frames from all input channels */
 			for (i = 0; i < 2 * in_nch; i++) {
 				cd->in[i] = *x << 16;
-				tdfb_direction_copy_emphasis(cd, in_nch, &emp_ch, *x << 16);
+				if (cd->direction_updates)
+					tdfb_direction_copy_emphasis(cd, in_nch, &emp_ch, *x << 16);
```

Comment:
this is a mystery - why does it help at all? The same check is done exactly as often as before. The only thing that it *potentially* saves is a repeated calculation of `*x << 16` which the compiler really should optimise away. Although maybe it cannot because it assumes, that the value can change since line 83? Can you try instead of this change to calculate `*x << 16` into a temporary variable and use it for both lines 83 and 84?

Reply:
@lyakh direction_copy_emphasis is in a different compilation unit (tdfb_direction.c), so there's a function call which surely is a cost.

Reply:
I'm usually trying to save code size, so that's why the check was there, but in this case I had missed that it had such big impact to performance.

Reply:
> @lyakh direction_copy_emphasis is in a different compilation unit (tdfb_direction.c), so there's a function call which surely is a cost.

@kv2019i really? Looking at the code that path is run in a loop of 2 * number of channels iterations, **after which** the actual data processing takes place which surely should take the lion portion of the time. And just avoiding a couple of `call*; entry; retw` instructions makes such a huge difference? Weird...

Reply:
@lyakh That's a different question how much is saved, but just noting there's a bigger difference than calculation of "*x << 16".

Reply:
@kv2019i that's likely true, thanks for the correction :-)